core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -214,71 +224,89 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
-        }
-
-        @Override
-        public BatchResult<ShardRoutingEntry> execute(ClusterState currentState, List<ShardRoutingEntry> tasks) throws Exception {
-            BatchResult.Builder<ShardRoutingEntry> batchResultBuilder = BatchResult.builder();
-
-            // partition tasks into those that correspond to shards
-            // that exist versus do not exist
-            Map<ValidationResult, List<ShardRoutingEntry>> partition =
-                tasks.stream().collect(Collectors.groupingBy(task -> validateTask(currentState, task)));
+        public BatchResult<ShardEntry> execute(ClusterState currentState, List<ShardEntry> tasks) throws Exception {
+            BatchResult.Builder<ShardEntry> batchResultBuilder = BatchResult.builder();
+            List<ShardEntry> validTasks = new ArrayList<>();
+            List<FailedRerouteAllocation.FailedShard> validShardRoutings = new ArrayList<>();
+            Set<ShardRouting> seenValidShardRoutings = new HashSet<>();
+
+            for (ShardEntry task : tasks) {
+                IndexMetaData indexMetaData = currentState.metaData().index(task.shardId.getIndex());
+                if (indexMetaData == null) {
+                    // tasks that correspond to non-existent shards are marked as successful
+                    logger.trace("{} index metadata for index {} does not exist anymore, marking task {} as successful", task.shardId,
+                        task.shardId.getIndex(), task);
+                    batchResultBuilder.success(task);
+                } else {
+                    // non-local requests
+                    if (task.primaryTerm > 0 && task.primaryTerm != indexMetaData.primaryTerm(task.shardId.id())) {
+                        logger.trace("{} primary term does not match for task {} (required [{}] but was [{}])", task.shardId,
+                            indexMetaData.primaryTerm(task.shardId.id()), task.primaryTerm, task);
+                        batchResultBuilder.failure(task, new NoLongerPrimaryShardException(
+                            task.shardId,
+                            "primary term [" + task.primaryTerm + "] did not match current primary term [" +
+                                indexMetaData.primaryTerm(task.shardId.id()) + "]"));
+                        continue;
+                    }
 
-            // tasks that correspond to non-existent shards are marked
-            // as successful
-            batchResultBuilder.successes(partition.getOrDefault(ValidationResult.SHARD_MISSING, Collections.emptyList()));
+                    ShardRouting matched = currentState.getRoutingTable().getByAllocationId(task.shardId, task.allocationId);
+                    if (matched == null) {
+                        // tasks that correspond to non-existent shards are marked as successful
+                        logger.trace("{} shard with allocation id {} does not exist anymore, marking task {} as successful", task.shardId,
```

Comment:
looking at the old code I see these are debug , and I tend to agree - how failures are handled is important (we long them on debug/warn on the other side). thoughts?